src/Processors/TTL/TTLAggregationAlgorithm.cpp

```diff
@@ -215,18 +217,26 @@ void TTLAggregationAlgorithm::finalizeAggregates(MutableColumns & result_columns
             for (const auto & it : description.set_parts)
                 it.expression->execute(agg_block);
 
+            std::unordered_set<String> columns_added;
             for (const auto & name : description.group_by_keys)
             {
                 const IColumn * values_column = agg_block.getByName(name).column.get();
                 auto & result_column = result_columns[header.getPositionByName(name)];
                 result_column->insertRangeFrom(*values_column, 0, agg_block.rows());
+                columns_added.emplace(name);
             }
 
             for (const auto & it : description.set_parts)
             {
                 const IColumn * values_column = agg_block.getByName(it.expression_result_column_name).column.get();
-                auto & result_column = result_columns[header.getPositionByName(it.column_name)];
-                result_column->insertRangeFrom(*values_column, 0, agg_block.rows());
+
+                /// In case the same column is used in the GROUP BY and in the SET clause, we need to make sure
+                /// it's only added once.
+                if (!columns_added.contains(it.column_name))
```

Comment:
I will be honest, this feels kind of wonky. Technically you are right, however it is not clear why the column is not updated. Have you considered:
1. Using the result of the `SET` expression instead of `GROUP BY`?
2. Have you considered preventing such cases when there are common expressions between `GROUP BY` and `SET`?

Reply:
>1. Using the result of the SET expression instead of GROUP BY?

It's still using the SET expression because of the prior

```cpp
            for (const auto & it : description.set_parts)
                it.expression->execute(agg_block);
```

>Have you considered preventing such cases when there are common expressions between GROUP BY and SET?

Yes, we discusses a little bit about it. I didn't want to break any existing tables. I've made my peace with this thing and even though I think using the same column in GROUP BY and in SET might be confusing, there might be use cases in which you may want to use it. I've updated the test in https://github.com/ClickHouse/ClickHouse/pull/82054/commits/c55601d3cd5a02ccd089e0178b0601a7527c4d9b to ensure it works as expected

Reply:
I've done a commit so that `SET` takes preference over `GROUP BY` because otherwise results are quite inconsistent: https://github.com/ClickHouse/ClickHouse/pull/82054/commits/a6fdb1063ffca82afc769b3eedb8fc2c57cb3ac6